### PR TITLE
Merge main into features/list-patterns

### PR DESCRIPTION
**Pull request by alrz**
```
# Conflicts:
#	src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
#	src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs
#	src/Compilers/CSharp/Portable/Errors/MessageID.cs
#	src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs
#	src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs
#	src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs
#	src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs
#	src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FunctionPointerInvocation.cs
#	src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs
#	src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs
#	src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.cs.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.de.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.es.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.fr.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.it.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.ja.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.ko.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.pl.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.pt-BR.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.ru.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.tr.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hans.xlf
#	src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hant.xlf
#	src/Compilers/Test/Utilities/CSharp/TestOptions.cs
```

### CODE REVIEW

**Comment by alrz** [comment]
@jcouv I've had already started attempting a merge. Feel free to close if you're half way through.

**Comment by jcouv** [comment]
Cool! There were more merge conflicts than I expected, so I'm glad you took a stab :-)

**Comment by alrz** [comment]
Most of it is a removed unused arg which I regret doing it in this branch. 
I listed conflicted files if you wanted to review.

**Comment by jcouv** [comment]
The only trouble is that we can't really review this PR as-is, as the merge conflicts are lost into the merge commit.
For situations where the merge isn't trivial (simple additive merges) we prefer to resolve all the conflicts in a separate commit which can be reviewed.

I think it's possible to re-do this with miminal effot:
- re-do a merge without resolving conflicts (ie. commit the conflict markers)
- use the last commit of this PR to update the files in your workspace to resolve conflicts again, commit that as the commit for resolving conflicts

**Comment by jcouv** [comment]
> I listed conflicted files if you wanted to review.

Okay. I think that works too. I'll take a look that way

**Comment by jcouv** [comment]
> I listed conflicted files if you wanted to review.

Actually that doesn't work :-/ GitHub won't let me jump to a specific file, because too many files changed...

**Comment by alrz** [comment]
I did a clean merge with "theirs" option to see the diff: https://github.com/dotnet/roslyn/compare/1cfbd6dc..81d9c767

Does that help?

**Comment by jcouv** [comment]
> Does that help?

I'm not sure.
What are the two commits in https://github.com/dotnet/roslyn/compare/1cfbd6dc..81d9c767 ? Neither appears in this PR.

I'd still recommend the approach I described earlier. It is tried and tested (one merge commit keeping conflicts, one commit to resolve conflicts).

**Comment by alrz** [comment]
>What are the two commits in https://github.com/dotnet/roslyn/compare/1cfbd6dc..81d9c767 ? Neither appears in this PR.

81d9c76 is the merge with manually resolved conflicts (I do see this in the PR)
1cfbd6dc is the clean merge from main with all conflicts resolved to main (by `-Xtheirs`)

**Comment by alrz** [comment]
>I'd still recommend the approach I described earlier. It is tried and tested (one merge commit keeping conflicts, one commit to resolve conflicts).

Done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -8051,7 +8051,7 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
                     argIsIndex
                         ? ErrorCode.ERR_ImplicitIndexIndexerWithName
                         : ErrorCode.ERR_ImplicitRangeIndexerWithName,
-                    arguments.Names[0].GetLocation());
+                    arguments.Names[0].GetValueOrDefault().Location);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:8054]: Is a `null` location what we want for the `OrDefault` case, as opposed to `Location.None`?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs [hunk 2]
```diff
@@ -326,14 +326,8 @@ private BoundExpression VisitAssignmentOperator(BoundAssignmentOperator node, bo
...
                 ref argumentRefKindsOpt,
-<<<<<<< HEAD
-                out argTemps,
-                invokedAsExtensionMethod: false);
-=======
                 ref argTempsBuilder,
-                invokedAsExtensionMethod: false,
-                enableCallerInfo: ThreeState.True);
->>>>>>> origin/main
+                invokedAsExtensionMethod: false);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs:330]: Why was `enableCallerInfo: ThreeState.True` removed?

File: src/Compilers/Test/Utilities/CSharp/TestOptions.cs [hunk 3]
```diff
@@ -56,11 +56,8 @@ public static class TestOptions
         public static readonly CSharpParseOptions RegularWithRecursivePatterns = Regular8;
         public static readonly CSharpParseOptions RegularWithoutPatternCombinators = Regular8;
         public static readonly CSharpParseOptions RegularWithPatternCombinators = RegularPreview;
-<<<<<<< HEAD
-        public static readonly CSharpParseOptions RegularWithListPatterns = RegularPreview;
-=======
         public static readonly CSharpParseOptions RegularWithExtendedPropertyPatterns = RegularPreview;
->>>>>>> origin/main
+        public static readonly CSharpParseOptions RegularWithListPatterns = RegularPreview;
```
Comment (jcouv) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs:60]: nit: this shouldn't be necessary anymore. The default version for all tests has been changed to `preview`.
Only LangVer tests will have to do something explicit (using RegularN for old and TemporaryPreview LangVer for new). TemporaryPreview will be replaced by RegularN+1 when the new language version is introduced.

**Review by jcouv — COMMENTED** [review]
Thanks much for restructuring the PR. It was much easier to review that way.
Looks good, just a question on possible null Location.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:8054]: This is a move. Changed in:

https://github.com/dotnet/roslyn/pull/52422/files#diff-94aa95e94cfdcc70e3ae5562a1fa1da77438bdb4c5499304caac8dc8e0133500R8015

File: src/Compilers/Test/Utilities/CSharp/TestOptions.cs (hunk 3 above)
Comment (alrz) [reply src/Compilers/Test/Utilities/CSharp/TestOptions.cs:60]: >The default version for all tests has been changed to preview.

Good to know. I can remove in some other PR (existing or upcoming) if that's ok.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:8054]: Thanks

**Review by jcouv — APPROVED** [review]

File: src/Compilers/Test/Utilities/CSharp/TestOptions.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/TestOptions.cs:60]: If you'd like. This is just something to know moving forward. Hopefully it should be a simplification.

**Pull request merged**